### ISSUE TITLE
parameterization: associated values

### ISSUE THREAD

**Issue by oliver-sanders** [cylc/cylc#2456]
A barrier to many users wishing to adopt parametrisation is the in-ability to associate values with the parameter as can be done using python data-structures via Jinja2 e.g:

```
{% set run_lengths = [
    {'name': 'short', 'time': 'PT10M', 'cpus', 4},
    {'name': 'short', 'time': 'PT20M', 'cpus', 8},
    {'name': 'short', 'time': 'PT1H30M', 'cpus', 64}
] %}

[runtime]
{% for run_length in run_lengths %}
    [[process_{{ run_length['name'] }}]]
        [[[job]]]
            execution time limit = {{ run_length['time'] }}
        [[[directives]]]
            cpus = {{ run_length['cpus'] }}
{% endfor %}

```
Might it be possible to replicate this functionality with parameterisations? i.e. something like:
```
[cylc]
    [[parameters]]
        run_length = """
                          time        cpus
            short:        PT10M       4
            medium:       PT20M       8
            long:         PT1H30M     64
        """
        model = """
                          id
            foo:          42
            bar:          24
        """

[runtime]
    [[process<model, run_length>]]
        [[[environment]]]
            MODEL_ID = <model.id>
        [[[job]]]
            execution time limit = <run_length.time>
        [[[directives]]]
            cpus = <run_length.cpus>
```
See also #2453.

**Comment by matthewrmshin** [comment]
Another use case is where users want to build a list of tasks using a multi-dimensional data structure. E.g.:

```
{% set seqs = {
    'prime': [2, 3, 5, 7, 11, 13],
    'fibonacci': [0, 1, 1, 2, 3, 5, 8, 13],
    'odd': [1, 3, 5, 7, 9, 11, 13],
} %}

[runtime]
{% for name, values in seqs.iteritems() %}
{% for value in values %}
    [[process_{{name}}_{{value}}]]
        # ...
{% endfor %}
{% endfor %}
```

The list in the 2nd dimension has variable length. (In real life, the 1st dimension may be the name of a physics section, and the 2nd dimension may be a list of diagnostic codes.)

**Comment by hjoliver** [comment]
Matt's use case with parameters (or am I missing something?):
```
[cylc]
  [[parameters]]
    prime = 2, 3, 5, 7, 11, 13
    fibonacci = 0, 1, 1, 2, 3, 5, 8, 13
    odd = 1, 3, 5, 7, 9, 11, 13
[runtime]
   [[process_<prime>, process_<fibonacci>, process_<odd>]]
        # ...
```
Note I could also define the parameters (names and values, if nec.) from your original Jinja2 data structure.  Any computation involving parameter under [runtime] has to be moved into task scripting using parameter value environment variables.

**Comment by matthewrmshin** [comment]
Yes, we can have some static in this example because dimension 1 is a small list of names, but the parameters in dimension 1 may be numbers, may need to be generated, or may be a long list. Or we may have even more dimensions.

**Comment by hjoliver** [comment]
But you can generate the parameter config section with Jinja2 loops over the data structure.

It would be good to have a concrete example that is impossible, or at least horrible, with parameters.

**Comment by matthewrmshin** [comment]
I think there is now a desire to use parameters where possible. It is just much neater than using Jinja2, especially in graphing. If we are able to use parameters for my example, we'd only need one liners for graph and a neat runtime section instead of multiple dimension of Jinja2 loops in multiple places.

**Comment by hjoliver** [comment]
Fair enough.  I've been countering some claims of "you can't do this with parameters" with the fact that you can use both parameters and Jinja2 at the same time (some users don't seem to know that), and that parameters were not intended to replace all use of Jinja2 loops, just loops over task names. However, I'm not averse to extending their use in this way.

**Comment by hjoliver** [comment]
But for the record, even Oliver's example above isn't particularly difficult right now - just parameterize over run-length - 'short', 'medium', 'long' - and have the task scripting select time and cpus according to value of the the parameter.  Not ideal because you have to give all values of time and cpus to all tasks, but still not that difficult or messy:
```
[runtime]
   [[process<run_len>]]
         # ...
         [[[environment]]]
               CPUS = {{ordered_cpu_list}}
               TIME = {{ordered_time_list}}
```

**Comment by arjclark** [comment]
> and have the task scripting select time and cpus according to value of the of the parameter

@hjoliver - The point of putting the timings, cpus, memory requests etc. in those has typically been for generation of directives, so the jinja using it would look something like what @oliver-sanders shows in the opening example. 

Perhaps I'm missing something from your response though? A better/more comprehensive example for the use case might be the sort of thing seen in the Global suite where there are different obs types with all their directives settings stored in lists/dicts (depending on which bit of the suite you look at). The suite just iterates over a list of obs to use and pulls in the entries from the obs dict to generate task directives. Something like:

```
{% for obs_type in OBS_TO_USE %}
[[ process_{{ obs_type }} ]]
inherit = OBS_FAM
  [[[ job ]]]
    execution time limit = {{ OBS_DICT[obs_type]['walltime'] }}
  [[[ directives ]]]
    -l nodes = {{ OBS_DICT[obs_type]['nodecount'] }}
{% endfor %}
```

(N.B. I've been generating parameter lists with Jinja2 loops and then using them from then on with some hybrid jinja2 to pick up on any of the gaps in parameterization. This dict type example means you can't really get away from the Jinja when it feels like it'd be nicer if something could be done with the parameterization)

Possible equivalent parameterization syntax for the above might look like:

```
[[ process<obs> ]]
inherit = OBS_FAM
  [[[ job ]]]
    execution time limit = <obs:walltime>
  [[[ directives ]]]
    -l nodes = <obs:nodecount>
```
